go-controller/pkg/ovn/policy.go

```diff
@@ -0,0 +1,532 @@
+
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1.Pod{},
+		time.Second*0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				oc.handleLocalPodSelectorAddFunc(policy, np, obj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				oc.handleLocalPodSelectorDelFunc(policy, np, obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oc.handleLocalPodSelectorAddFunc(policy, np, newObj)
+			},
+		},
+	)
+	channel := make(chan struct{})
+	go controller.Run(channel)
+	select {
+	case <-np.stop:
+		channel <- struct{}{}
+	}
+}
+
+func (oc *Controller) handlePeerPodSelector(
+	policy *kapisnetworking.NetworkPolicy, podSelector *metav1.LabelSelector,
+	address_set string, stop chan bool) {
+
+	address_map := make(map[string]bool)
+
+	client, _ := oc.Kube.(*kube.Kube)
+	clientset, _ := client.KClient.(*kubernetes.Clientset)
+	podSelectorAsSelector, _ := metav1.LabelSelectorAsSelector(podSelector)
+
+	watchlist := newListWatchFromClient(clientset.Core().RESTClient(), "pods",
+		policy.Namespace, podSelectorAsSelector)
+
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1.Pod{},
+		time.Second*0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				pod := obj.(*kapi.Pod)
+				if pod.Status.PodIP == "" ||
+					address_map[pod.Status.PodIP] == true {
+					return
+				}
+				address_map[pod.Status.PodIP] = true
+				addresses := make([]string, 0, len(address_map))
+				for k := range address_map {
+					addresses = append(addresses, k)
+				}
+				oc.setAddressSet(address_set, addresses)
+			},
+			DeleteFunc: func(obj interface{}) {
+				pod := obj.(*kapi.Pod)
+
+				ip_address := oc.getIpFromOvnAnnotation(pod.Annotations["ovn"])
+				if ip_address == "" {
+					return
+				}
+
+				if address_map[ip_address] == false {
+					return
+				}
+
+				delete(address_map, ip_address)
+
+				addresses := make([]string, 0, len(address_map))
+				for k := range address_map {
+					addresses = append(addresses, k)
+				}
+				oc.setAddressSet(address_set, addresses)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				pod := newObj.(*kapi.Pod)
+				if pod.Status.PodIP == "" ||
+					address_map[pod.Status.PodIP] == true {
+					return
+				}
+				address_map[pod.Status.PodIP] = true
+				addresses := make([]string, 0, len(address_map))
+				for k := range address_map {
+					addresses = append(addresses, k)
+				}
+				oc.setAddressSet(address_set, addresses)
+			},
+		},
+	)
+	channel := make(chan struct{})
+	go controller.Run(channel)
+	select {
+	case <-stop:
+		channel <- struct{}{}
+	}
+}
+
+func (oc *Controller) addNetworkPolicy(policy *kapisnetworking.NetworkPolicy) {
+	logrus.Infof("Adding network policy %s in namespace %s", policy.Name,
+		policy.Namespace)
+
+	if oc.namespacePolicies[policy.Namespace] != nil &&
+		oc.namespacePolicies[policy.Namespace][policy.Name] != nil {
+		return
+	}
+
+	oc.waitForNamespaceEvent(policy.Namespace)
```

Comment:
We probably should get rid of this function call, and make the network policy map in-memory anyway, irrespective of namespace create/update.